src/EditorFeatures/Core.Wpf/InlineDiagnostics/InlineDiagnosticsAdornmentManager.cs

```diff
@@ -0,0 +1,214 @@
+                    // each line with errors will be ordered by importance
+                    if (value.mapTagSpan is not null && value.mapTagSpan.Tag.ErrorType is not PredefinedErrorTypeNames.SyntaxError)
+                    {
+                        map[lineNum] = (tagMappingSpan, changedSpan);
+                    }
+                }
+            }
+
+            return map;
+        }
+
+        /// <summary>
+        /// Iterates through the mapping of line number to span and draws the diagnostic in the appropriate position on the screen,
+        /// as well as adding the tag to the adornment layer.
+        /// </summary>
+        protected override void AddAdornmentsToAdornmentLayer_CallOnlyOnUIThread(NormalizedSnapshotSpanCollection changedSpanCollection)
+        {
+            // this method should only run on UI thread as we do WPF here.
+            Contract.ThrowIfFalse(TextView.VisualElement.Dispatcher.CheckAccess());
+
+            var viewLines = TextView.TextViewLines;
+            var map = GetSpansOnEachLine(changedSpanCollection);
+            foreach (var (lineNum, (tagMappingSpan, snapshotSpan)) in map)
+            {
+                if (!TryMapToSingleSnapshotSpan(tagMappingSpan.Span, TextView.TextSnapshot, out var span))
+                {
+                    continue;
+                }
+
+                var geometry = viewLines.GetMarkerGeometry(span);
+                if (geometry != null)
+                {
+                    var tag = tagMappingSpan.Tag;
+                    var classificationType = _classificationRegistryService.GetClassificationType(InlineDiagnosticsTag.TagID + tag.ErrorType);
+                    var graphicsResult = tag.GetGraphics(TextView, geometry, GetFormat(classificationType));
+
+                    // Need to get the SnapshotPoint to be able to get the IWpfTextViewLine
+                    var point = tagMappingSpan.Span.Start.GetPoint(snapshotSpan.Snapshot, PositionAffinity.Predecessor);
+                    if (point == null)
+                    {
+                        continue;
+                    }
+
+                    var lineView = TextView.GetTextViewLineContainingBufferPosition(point.Value);
+
+                    var visualElement = graphicsResult.VisualElement;
+                    if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfCode)
+                    {
+                        Canvas.SetLeft(visualElement, lineView.Right);
+                    }
+                    else if (tag.Location is InlineDiagnosticsLocations.PlacedAtEndOfEditor)
+                    {
+                        Canvas.SetLeft(visualElement, TextView.ViewportWidth - visualElement.DesiredSize.Width);
+                    }
+
+                    Canvas.SetTop(visualElement, geometry.Bounds.Bottom - visualElement.DesiredSize.Height);
+
+                    // Only place the diagnostics if the diagnostic would not intersect with the editor window
```

Comment:
this is probably good to start.  part of me thinks we should investigate drawing this, but cutting off on the RHS of the editor.    Then as the user scrolls right, they'd see more of the error come into view.  that can be done outside of this PR.

Reply:
That sounds like a good idea, I can create mock ups for that.